File: examples/python/javacount.py
import logging
import re
import typing

import apache_beam as beam
from apache_beam.io import ReadFromText
from apache_beam.io import WriteToText
from apache_beam.transforms.external import ImplicitSchemaPayloadBuilder
from apache_beam.options.pipeline_options import PipelineOptions


class WordExtractingDoFn(beam.DoFn):
  """Parse each line of input text into words."""
  def process(self, element):
    """Returns an iterator over the words of this element.
    The element is a line of text.  If the line is blank, note that, too.
    Args:
      element: the element being processed
    Returns:
      The processed element.
    """
    return re.findall(r'[\w\']+', element, re.UNICODE)


def run(input_path, output_path, pipeline_args):
  pipeline_options = PipelineOptions(pipeline_args)

  with beam.Pipeline(options=pipeline_options) as p:
    lines = p | 'Read' >> ReadFromText(input_path).with_output_types(str)
    words = lines | 'Split' >> (beam.ParDo(WordExtractingDoFn()).with_output_types(str))

    java_output = (
        words
        | 'JavaCount' >> beam.ExternalTransform(
              'my.beam.transform.javacount',
              None,
              "localhost:12345"))

    def format(kv):
      key, value = kv
      return '%s:%s' % (key, value)

    output = java_output | 'Format' >> beam.Map(format)
    output | 'Write' >> WriteToText(output_path)


if __name__ == '__main__':
  logging.getLogger().setLevel(logging.INFO)
  import argparse

  parser = argparse.ArgumentParser()
  parser.add_argument(
      '--input',
      dest='input',
      required=True,
      help='Input file')
  parser.add_argument(
      '--output',
      dest='output',
      required=True,
      help='Output file')
  known_args, pipeline_args = parser.parse_known_args()

  run(
      known_args.input,
      known_args.output,
      pipeline_args)

File: examples/java/src/main/java/my/beam/transforms/JavaCount.java
package my.beam.transforms;

import org.apache.beam.sdk.transforms.Count;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.PTransform;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;

public class JavaCount extends PTransform<PCollection<String>, PCollection<KV<String, Long>>> {

  public JavaCount() {
  }

  @Override
  public PCollection<KV<String, Long>> expand(PCollection<String> input) {
    return input
        .apply(
            "JavaCount", Count.perElement());
  }
}
